### ISSUE TITLE
plain TCP RST packet in pod-to-pod-with-l7-policy-encryption for IPv6+IPSec+VXLan

### ISSUE THREAD

**Issue by smagnani96** [cilium/cilium#35485]
After fixing the MTU issue when having IPSec with both in+egress proxy in #35173, we enabled the `pod-to-pod-with-l7-policy-encryption` test for IPv4. When trying to enable also the test for IPv6, we see a flaky behavior:

1. native routing: after the MTU patch, no problems are encountered and the tests execute smoothly.
2. overlay mode: either with or without the MTU patch, enabling the tests would cause a plain-test TCP RST packet to be leaked when using VxLan as overlay. This behavior is probably unrelated to the MTU patch and should be related only to CI and not to the whole IPv6 functioning. A guessing would be that bpftrace is still monitoring looking for plain-text packets while tearing down Cilium and something's probably happening in between.

As discussed in the previous MTU issue, the failing tests seems randomly happen with kernels `rhel8-20240404.144247`, `5.15-20240710.064909`, and  `bpf-next-20240711.013133`. Testing with Geneve is ok, while with VXLan there's this problem.

**Comment by smagnani96** [comment]
From a 1st analysis in the previous issue, it seemed that this flake was constant, always occurring upon every run.
I wanted to check if I was able to reproduce locally with no luck. Every time I run the test it passes, and the bpftrace script does not detect any plain-text packet. BUT, I decided to keep monitoring the two nodes with client-server pods for a while ever after the test finished with success.

What I noticed, is that after a random amount of time (still close to ~1min), a multitude of TCP FYN-ACK packets are recorded in the receiver node. After more 4min, there's also a TCP RST packet recorded in the receiver node as well. 
RST packet is being dropped with `SKB_DROP_REASON_TCP_RESET`. The others are dropped with `SKB_DROP_REASON_NO_SOCKET`, and `SKB_DROP_REASON_NOT_SPECIFIED`. The 1st drop ever registered is `SKB_DROP_REASON_QUEUE_PURGE`.

I'm still investigating this behavior and think whether we could hit the same in CI, even if there are basically no waiting periods between these packets (RST recorded as soon as the test finishes, no need to wait).

In the meantime, I attach to this both the pcaps and pwru output.
[issue.zip](https://github.com/user-attachments/files/17981880/issue.zip)

**Comment by smagnani96** [comment]
Same happened in https://github.com/cilium/cilium/pull/36364#issuecomment-2517956205 where the `pod-to-pod-with-l7-policy-encryption` test is not enabled yet.

We believe that the 5min delay of the RST packet is due to the Envoy [stream_idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#:~:text=to%20around%201024kb.-,stream_idle_timeout,while%20introducing%20robustness%20to%20TCP%20connections%20that%20terminate%20without%20a%20FIN.,-This%20idle%20timeout) parameter: in this case, while the client has already closed the connection (SKB_DROP_REASON_NO_SOCKET), the envoy does not immediately close connection in case it can reuse it, and after the 5min it generates the RST packet.

Differently than https://github.com/cilium/cilium/issues/33084, in the current issue the bpftrace output reports `override: 0`, meaning that the packet does not come from a proxy 🤔

I'm quite sure this RST packet comes from a previous test being run, for which the envoy connection will ~5min later generate the RST packet. At this point, we should still be robust against previous flying packets, but maybe loading/removing policies or flushing ct in the meantime causes problems.

Potentially related https://github.com/cilium/cilium/issues/31651:packets being mistakenly forwarded when using L7 and CiliumInternalIP when no active socket.

**Comment by giorio94** [comment]
> Differently than https://github.com/cilium/cilium/issues/33084, in the current issue the bpftrace output reports override: 0, meaning that the packet does not come from a proxy 🤔

Related to this, consider that the script [stops tracking](https://github.com/cilium/cilium/blob/7b5023018fb97233a36cfd7af03da20110a6db09/.github/actions/bpftrace/scripts/check-ipsec-leaks.bt#L190-L202) a given source IP/port pair once the TCP connection is closed. That may explain `override=0`, even if the traffic was supposed to come from a proxy, considering that the connection is not closed cleanly here.

**Comment by giorio94** [comment]
> In fact, the connection that I'm seeing even with override=0 are still node->pod

Are you referring to the one mentioned in https://github.com/cilium/cilium/pull/36364#issuecomment-2517956205? If that's the case, then it is actually from the ingress endpoint (a special endpoint representing the proxy in case of ingress/gateway API traffic) to pod. So yes, I would normally expect it to be flagged as coming from the proxy.

```
    ingress:
      ipv4: 10.244.2.71
      ipv6: fd00:10:244:2::8dcc
```

And yes, that looks definitely related to the behavior I had observed in https://github.com/cilium/cilium/issues/31651 (I just updated that issue to correctly mention the ingress IP, rather than the Cilium Internal IP. Both from memory and re-reading the description --- the Cilium Internal IP is actually assigned to the `cilium_host` interface --- I was definitely referring to the former, although I had mistakenly written the latter).

**Comment by jschwinger233** [comment]
I managed to reproduce in lvh, but failed to simplify the steps and reduce repro time: it has to run the full conn test.

steps:

```
lvh images pull --cache quay.io/lvh-images/kind:bpf-next-20240711.013133 --dir ./
lvh run --host-mount=./ --image ./images/kind_bpf-next.qcow2       --daemonize -p 2222:22 --serial-port 0       --cpu=4 --mem=12G --cpu-kind host       --console-log-file /tmp/console.log

ssh -p 2222 -o "StrictHostKeyChecking=no" root@localhost
# then inside the lvh
(lvh)

**Comment by smagnani96** [comment]
Many thanks for the effort 🙏 

> ## TLDR
> My theory is, L7 proxy (envoy) can't set proper mark to tcp reset. I'm not sure if this is a kernel limitation or our userspace process wasn't programmed correctly, I'll dig deeper next Monday.

Before holidays I also managed to reproduce this, but didn't fully understand the root cause.
As mentioned before, it seems like our L7 proxy sends keeps connections alive and sends a reset packet only after a given amount of time. That is why running the single cli test pod-to-pod-with-l7-encryption doesn't lead to leaks.

> As for where tcp reset came from, I don't know yet, but I presume it's ordinary (and inevitable?) to see tcp reset segments in production, due to many reasons. The key question might be: should cilium transparent encryption also takes care of tcp reset; if yes, how?

Here I don't fully understand yet why they don't get marked.
If they were marked, I'd expect us to handle them as all the other proxy-to-pod packets -> https://github.com/cilium/cilium/blob/main/bpf/bpf_host.c#L885.

> ## Repro

Same steps as yours, using lvh and bpf-next as kernel to reproduce.
Repro seems to be consistent among multiple runs.
Running individual test (pod-to-pod-with-l7-policy-encryption) passes with no leaked packet.
In IPv4-only test I never see the plain-text reset packet.

> ## Investigation

Here I also attach the sysdump I generated, as well as my pwru trace until the moment of the leak (whole cli tests) and bpftrace [leak-sysdump.zip](https://github.com/user-attachments/files/18273616/leak-sysdump.zip), [leak_pwru.log](https://github.com/user-attachments/files/18273617/leak_pwru.log), [leak_bpftrace.log](https://github.com/user-attachments/files/18273618/leak_bpftrace.log).

> Outstanding issues:
> 
> 2. Is this a ipv6 only issue, or ipv4 is also affected?

For IPv4 I've never seen it nor in CI nor locally. Running v4-only tests doesn't cause the leak. 
  
> 4. Is it possible to handle these tcp reset. I mean, our datapath should after all encrypt them? (VinE helps?)

I don't see why we shouldn't try to encrypt those packets. Theoretically, they are proxy-to-pod packets (even though we don't have the mark set yet), therefore imho they should be encrypted as well. VinE might help to avoid leaking a plain-text packet (despite being an empty reset), but I'm not sure how the receiver side would handle it.
  
> 4. Why were these tcp reset segments generated?

I think a connections is kept open by the L7 proxy and after a certain amount of time it tries to close it.
Still trying to narrow down the exact test in which they were originally generated though.

**Comment by jschwinger233** [comment]
I followed the path Simone pointed to, got a new theory:

For a tcp socket with non-zero linger (e.g. `setsockopt(SOL_SOCKET, SO_LINGER, {l_onoff=1, l_linger=10})`) and mark (e.g. `setsockopt(SOL_SOCKET, SO_MARK, 0xa00)`), if close(socket) when there is unsent buffer in socket (e.g. packets sent by this socket don't get ack), close() syscall blocks for 10 seconds (linger=10), followed by a TCP reset.

This TCP reset may not have 0xa00 mark?

This comes from envoy's long connection pointed out by Simone: even after `cilium-cli connectivity --test pod-to-pod-encryption` completes, we can still see the sockets using `lsof`:

```
$ nspod cilium-envoy-sbcb7 lsof -Nnp 276201 

cilium-en 276201 root  198u     IPv4            3027808      0t0      TCP 10.244.2.254:http-alt->10.244.3.48:33428 (ESTABLISHED)
cilium-en 276201 root  199u     unix 0xffff9dce2f9f7400      0t0  3031685 type=SEQPACKET
cilium-en 276201 root  202u     IPv6            3028914      0t0      TCP [fd00:10:244:2::60ac]:http-alt->[fd00:10:244:3::2b3a]:56374 (ESTABLISHED)
```

So I `strace`d envoy process, to see if there is any special settings for those sockets:

```
$ strace -fTttp $envoy

[pid 277462] 17:41:34.331240 accept4(196, {sa_family=AF_INET6, sin6_port=htons(56374), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "fd00:10:244:3::2b3a", &sin6_addr), sin6_scope_id=0}, [128 => 28], SOCK_NONBLOCK) = 202 <0.000068>
[pid 277462] 17:41:34.331436 getsockopt(202, SOL_TCP, TCP_INFO, "\1\0\0\0\0\7w\1(\21\3\0@\234\0\0\25\5\0\0\30\2\0\0\0\0\0\0\0\0\0\0"..., [104]) = 0 <0.000031>
[pid 277462] 17:41:34.331553 epoll_ctl(55, EPOLL_CTL_ADD, 202, {events=EPOLLIN|EPOLLRDHUP|EPOLLET, data={u32=202, u64=202}}) = 0 <0.000018>
[pid 277462] 17:41:34.331745 epoll_wait(55, [{events=EPOLLIN, data={u32=202, u64=202}}], 32, 0) = 1 <0.000008>
[pid 277462] 17:41:34.331778 recvfrom(202, "GET / HTTP/1.1\r\nhost: [fd00:10:2"..., 65536, MSG_PEEK, NULL, NULL) = 254 <0.000016>
[pid 277462] 17:41:34.331840 getsockname(202, {sa_family=AF_INET6, sin6_port=htons(8080), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "fd00:10:244:2::60ac", &sin6_addr), sin6_scope_id=0}, [128 => 28]) = 0 <0.000008>
[pid 277462] 17:41:34.332745 setsockopt(202, SOL_SOCKET, SO_LINGER, {l_onoff=1, l_linger=10}, 8) = 0 <0.000011>
[pid 277462] 17:41:34.332776 setsockopt(202, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0 <0.000010>
[pid 277462] 17:41:34.332806 setsockopt(202, SOL_TCP, TCP_KEEPINTVL, [300], 4) = 0 <0.000009>
[pid 277462] 17:41:34.332834 setsockopt(202, SOL_TCP, TCP_KEEPIDLE, [300], 4) = 0 <0.000009>
[pid 277462] 17:41:34.332866 epoll_ctl(55, EPOLL_CTL_DEL, 202, 0x7ba2c87fe434) = 0 <0.000012>
[pid 277462] 17:41:34.332921 epoll_ctl(55, EPOLL_CTL_ADD, 202, {events=EPOLLIN|EPOLLOUT|EPOLLET, data={u32=202, u64=202}}) = 0 <0.000013>
[pid 277462] 17:41:34.332990 setsockopt(202, SOL_TCP, TCP_NODELAY, [1], 4) = 0 <0.000010>
[pid 277462] 17:41:34.333026 epoll_wait(55, [{events=EPOLLIN|EPOLLOUT, data={u32=202, u64=202}}], 32, 79) = 1 <0.000011>
[pid 277462] 17:41:34.333063 readv(202, [{iov_base="GET / HTTP/1.1\r\nhost: [fd00:10:2"..., iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}, {iov_base="", iov_len=16384}], 8) = 254 <0.000022>
[pid 277462] 17:41:34.333113 readv(202, [{iov_base=0x52b3f40c0fe, iov_len=16130}, {iov_base=0x52b3f410000, iov_len=16384}, {iov_base=0x52b3f414000, iov_len=16384}, {iov_base=0x52b3f418000, iov_len=16384}, {iov_base=0x52b3f41c000, iov_len=16384}, {iov_base=0x52b3f420000, iov_len=16384}, {iov_base=0x52b3f424000, iov_len=16384}, {iov_base=0x52b3f428000, iov_len=16384}], 8) = -1 EAGAIN (Resource temporarily unavailable) <0.000012>
[pid 277462] 17:41:34.337620 writev(202, [{iov_base="HTTP/1.1 200 OK\r\nx-powered-by: E"..., iov_len=390}, {iov_base="<html>\n  <head>\n    <link\n      "..., iov_len=2057}], 2) = 2447 <0.001793>
[pid 277462] 17:51:34.956773 epoll_ctl(55, EPOLL_CTL_DEL, 202, 0x7ba2c87fdeb4 <unfinished ...>
[pid 277462] 17:51:34.957013 close(202 <unfinished ...>
[pid 277445] 17:51:45.120401 <... close resumed>) = 0 <10.165733>
```

We can see the socket has kept there for another 10min before closing, and I want to highlight the last two lines:

```
[pid 277462] 17:51:34.957013 close(202 <unfinished ...>
[pid 277445] 17:51:45.120401 <... close resumed>) = 0 <10.165733>
```

close(fd) cost 10 seconds because of linger=10, and this indicates there was non-ack-ed data in socket buffer, which could end up with a TCP reset after 10 seconds waiting.

I suspect this TCP reset doesn't have the proper mark set by setsockopt().

**Comment by jschwinger233** [comment]
I might figure it out 😺 let me have a discussion with Simone first

**Comment by jschwinger233** [comment]
TL;DR

The plain-text tcp-rst was sent by kernel as the response to tcp-fin sent by envoy after idle timeout. Kernel associates this skb to envoy so we wrongly thought the tcp-rst was sent by envoy. The tcp-rst wasn't linked to envoy socket, so it didn't have the mark.

The question I'll ask is, why kernel responded envoy's tcp-fin with tcp-rst. 

Reproduce

I'll use a local ipv6 tcp echo server to demonstrate the idea.

1. run this ipv6 echo server. the listener socket is set to SO_MARK=0xa00, so traffic from listener and client sockets will have 0xa00 mark: https://github.com/jschwinger233/proving-ground/blob/main/go_tcp_echo_mark/main.go
2. use nc to connect and send some bytes to make sure echo server is working
```
$ nc ::1 8080
asdfasfdasdfa
asdfasfdasdfa
```
3. use ss to see connection
```
$ ss -nt | grep 8080
ESTAB      0      0                       [::1]:53500                   [::1]:8080        
ESTAB      0      0                       [::1]:8080                    [::1]:53500     
```

4. use ip6tables to drop traffic to 8080
```
sudo ip6tables -I OUTPUT 1 -o lo  -p tcp --dport 8080 -j DROP
```

5. kill the nc, wait a while until the fin-wait-1 socket is gone, when we'll see the solo established socket on server side
```
$ ss -nt | grep 8080
ESTAB      0      0                       [::1]:8080                    [::1]:53500
```

6. delete the ip6tables rule
```
sudo ip6tables -D OUTPUT -o lo  -p tcp --dport 8080 -j DROP
```

7. run pwru to capture tcp-rst
```
pwru --output-caller --filter-track-skb 'ip6 and port 53500 and ip6[53]&0x4!=0'
```

8. hit ctrl-c from echo server terminal to trigger socket close and tcp-fin from server. This leads to tcp-rst reply from kernel, although it looks like from echo server
```
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 0          0               0         0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) nf_hook_slow                 ip6_xmit
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 0          0               0         0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) ip6_output                   ip6_xmit
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) nf_hook_slow                 ip6_output
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) apparmor_ip_postroute        nf_hook_slow
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) ip6_finish_output            ip6_output
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 60    [::1]:53500->[::1]:8080(tcp) ip6_finish_output2           ip6_finish_output
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 74    [::1]:53500->[::1]:8080(tcp) __dev_queue_xmit             neigh_hh_output
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 74    [::1]:53500->[::1]:8080(tcp) tc_run                       __dev_queue_xmit
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 74    [::1]:53500->[::1]:8080(tcp) tcf_classify                 tc_run
0xffff9f5fa94c5c00 8   ~ho_mark:1076590 4026531840 0              lo:1       0x86dd 65536 74    [::1]:53500->[::1]:8080(tcp) skb_ensure_writable          bpf_skb_pull_data
```

from above result, we see this tcp-rst:
1. doesn't have 0xa00 mark
2. looks like from server process (process name is echo_mark)
3. traffic tuple is from client to server

All the characteristics match what we have observed before.

---

Simone asked what if the tcp-rst was actually sent by envoy instead of the reply from kernel, but I don't think that's the case.

I did some experiments and found, once the listener socket is set SO_MARK=0xa00, the accepted client socket will mark everything from it even if it's tcp-rst. I tested by `setsockopt(sock, linger, 0)` followed by `close(sock)`, so socket can sent tcp-rst on close.

---

What's still unclear to me is why kernel responds tcp-fin with tcp-rst. Simone suspects this is because cilium policy on and off frequently during connectivity test makes some tcp-fin dropped, leaving the envoy's long connection sockets lingering. When envoy's idle timeout expires, the tcp-fin from envoy will end up with tcp-rst from kernel, and that's the plain-text tcp-rst caught by leak detection.

**Comment by smagnani96** [comment]
Apologize for the bad write-up, will improve it in the afternoon.

---

We're close. The timeout intuition is correct, and Gray's experiments were amazingly useful to continue sorting this issue out. Let me continue from:

> What's still unclear to me is why kernel responds tcp-fin with tcp-rst. Simone suspects this is because cilium policy on and off frequently during connectivity test makes some tcp-fin dropped, leaving the envoy's long connection sockets lingering. When envoy's idle timeout expires, the tcp-fin from envoy will end up with tcp-rst from kernel, and that's the plain-text tcp-rst caught by leak detection.

I was able to identify all the network packets involved and our target leaked RST packet.
However, to observe the leak you might need the whole test suite (or most of it): by running only the `pod-to-pod-with-l7-policy-encryption` test I was able to identify all the packets, including the then leaked RST one, but in this case the bpftrace script didn't record it. I believe this has something to do with the current network policies being set at the time this RST is generated that cause the RST to being mistakenly forwarded when running the long test suite.

Steps to repro on lvh:

1. `docker run --network host -v /root/.kube/config:/root/.kube/config -v /host:/root/app quay.io/cilium/cilium-cli-ci:2d41ff7d68b7a1843fe00722cec4ec3c374c89f0 cilium connectivity test --include-unsafe-tests --sysdump-hubble-flows-count=1000000 --sysdump-hubble-flows-timeout=5m --flush-ct -v --timestamp  --test pod-to-pod-with-l7-policy-encryption`: run connectivity tests with only the `pod-to-pod-with-l7-policy-encryption`.
2. [Optional] `sudo /host/pwru --output-meta --filter-track-skb --output-caller --output-tuple 'ip6 and (src port 8080 or dst port 8080)' --timestamp absolute`: run pwru outputting tuple, tracking skb and filtering by IPv6 with either src or dst port 8080. I used it just to verify that the last RST has not the mark set.
3. `bpftrace /host/trace_drops.bt`: script to track skb dropped packets with either src or dst port 8080, printing tuples, timestamp and drop reason. 
4. `tcpdump -l -n -i any 'ip6 and (src port 8080 or dst port 8080)'`: record IPv6 packets for an 8080 port on any interface. Did this on both the two nodes (client-server)

[investigation.zip](https://github.com/user-attachments/files/18371616/investigation.zip)


A difference between tcpdump output captured in client node vs server node:

```log
05:59:54.379607 lxc312b400946c9 Out IP6 fd00:10:244:1::e942.8080 > fd00:10:244:3::2adc.57776: Flags [P.], seq 4294766024:4294768472, ack 4294776458, win 502, options [nop,nop,TS val 738216038 ecr 1437162223], length 2448: HTTP: HTTP/1.1 200 OK
05:59:54.379885 lxc312b400946c9 In  IP6 fd00:10:244:3::2adc.57776 > fd00:10:244:1::e942.8080: Flags [.], ack 4294768472, win 502, options [nop,nop,TS val 1437162275 ecr 738216038], length 0
05:59:54.383869 lxc312b400946c9 In  IP6 fd00:10:244:3::2adc.57776 > fd00:10:244:1::e942.8080: Flags [F.], seq 4294776458, ack 4294768472, win 502, options [nop,nop,TS val 1437162279 ecr 738216038], length 0
05:59:54.389742 lxc312b400946c9 Out IP6 fd00:10:244:1::e942.8080 > fd00:10:244:3::2adc.57776: Flags [F.], seq 4294768472, ack 4294776459, win 502, options [nop,nop,TS val 738216048 ecr 1437162279], length 0
05:59:54.389956 lxc312b400946c9 In  IP6 fd00:10:244:3::2adc.57776 > fd00:10:244:1::e942.8080: Flags [.], ack 4294768473, win 502, options [nop,nop,TS val 1437162285 ecr 738216048], length 0
```
Unique connection `pod->pod`, while in the node with the echo server there are two (see tcpdump file):

1. fd00:10:244:1::36d1.57092 > fd00:10:244:1::e942.8080: `router->pod` gets correctly terminated.
1. fd00:10:244:3::2adc.57776 > fd00:10:244:1::e942.8080: `pod->pod` does not terminate and causes the subsequent timeouts to run. But which packet did we loose?

TODO: I'll come back with considerations on this this afternoon. The zip should contain also the list of dropped packets.

---

There are multiple time-based events (the one that causes the leakage is the n.2):

1. after 1min and 2min: client FIN but server RST. Possible timer we use is (i) envoy.idleTimeoutDurationSeconds: Set Envoy upstream HTTP idle connection timeout seconds. Does not apply to connections with pending requests. Default 60s
2. after 5min and 10min: server FIN but client RST. Possible timers [TCP idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/_sources/faq/configuration/timeouts.rst.txt#:~:text=*%20The%20TCP%20protocol,idle_timeout%3E%60%20to%200.) (default set to 10min) and [HTTP stream_idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/_sources/faq/configuration/timeouts.rst.txt#:~:text=The%20HTTP%20connection%20manager%20%3Aref%3A%60stream_idle_timeout,be%20sent%20to%20a%20downstream%20client.) (default set to 5min).

I believe we're not correctly setting all timeout correctly, or that in case they're buggy for IPv6 in Envoy. For instance, I see us using `stream_idle_timeout=0` but there's still an event after exact 5min. The `TCP idle_timeout` set at 10min I might understand, since if we're missing a FIN-ACK packet the socket doesn't get closed before the time.

---

I Found old envoy bugs potentially related with timers/unilateral FINs, but I need to inspect them further:

1. https://github.com/envoyproxy/envoy/issues/6815
2. https://github.com/envoyproxy/envoy/issues/6190
3. https://github.com/envoyproxy/envoy/pull/6578

**Comment by jschwinger233** [comment]
Thank you Simone. Julian suggests we make leak detection script tolerable to plain-text tcp rst. If no huge efforts are required, it's better only the specific plain-text tcp rst related to envoy proxy will be tolerated.

This is because envoy traffic ending up with tcp rst reply is, I wouldn't say expected, but ordinary in the real world with complicated network settings. For now, let's tolerate these sort of leaks.

In future, I suppose things can improve as long as VinE lands: at that time, plain-text rst will reach ipsec hooks at to-netdev, where we can probably fix the leakage. We can review the issue then.

**Comment by dylandreimerink** [comment]
Now that this issue is resolved, can we also enable the pod to pod encryption connectivity tests? They are currently still guarded, referencing this issue as the blocker https://github.com/cilium/cilium/commit/ca1dec2f3d699b822ad8a343be4214f9374a3c13.

**Comment by smagnani96** [comment]
> Now that this issue is resolved, can we also enable the IPv6 pod to pod encryption connectivity tests? They are currently still guarded, referencing this issue as the blocker [ca1dec2](https://github.com/cilium/cilium/commit/ca1dec2f3d699b822ad8a343be4214f9374a3c13).

Wish we could, I'm afraid there's something new that needs to be addressed first https://github.com/cilium/cilium/issues/37051.
